ocaml/xapi/xapi_xenops.ml

```diff
@@ -1039,29 +1039,41 @@ module Xapi_cache = struct
       		updated whenever we receive an event from xapi. *)
 
   let cache = Hashtbl.create 10 (* indexed by Vm.id *)
+  let mutex = Mutex.create ()
+  let with_lock f =
+    Mutex.execute mutex f
+
+  let register id initial_value =
+    with_lock (fun () ->
+        debug "xapi_cache: creating cache for %s" id;
+        if not(Hashtbl.mem cache id) || (Hashtbl.find cache id = None)
+        then Hashtbl.replace cache id initial_value)
+
+  let unregister id =
+    with_lock (fun () ->
+      debug "xapi_cache: deleting cache for %s" id;
+      Hashtbl.remove cache id)
+
+  let update id f =
+    let find_nolock id =
+      if Hashtbl.mem cache id
```

Comment:
Same as above. If you don't like exceptions, you can use `Hashtbl.find_opt`

Reply:
Good point, when the original code was written find_opt did not exist, but we can use it now. I'll update.